### PR TITLE
feat(sdk): Add ClientReports

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

If clientReport exists it will be send with the next envelope.

## :bulb: Motivation and Context
Closes: https://github.com/getsentry/sentry-react-native/issues/2034

## :green_heart: How did you test it?

Unit tests, sample app, checked the client dropped in the graph

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: src/js/sdk.tsx [hunk 1]
```diff
@@ -41,6 +41,7 @@ const DEFAULT_OPTIONS: ReactNativeOptions = {
   transportOptions: {
     textEncoder: makeUtf8TextEncoder(),
   },
+  sendClientReports: true,
```
Comment (marandaneto) [inline src/js/sdk.tsx:44]: Is it disabled by default on JS?

Comment (krystofwoldrich) [reply src/js/sdk.tsx:44]: no

File: src/js/client.ts [hunk 2]
```diff
@@ -144,4 +158,23 @@ export class ReactNativeClient extends BaseClient<ReactNativeClientOptions> {
...
+    if (outcomes.length > 0) {
+      const clientReportItem: ClientReportItem = [
+        { type: 'client_report' },
+        {
+          timestamp: dateTimestampInSeconds(),
+          discarded_events: outcomes,
+        },
+      ];
+
+      envelope[items].push(clientReportItem);
```
Comment (marandaneto) [inline src/js/client.ts:177]: What if the envelope already has a `clientReportItem` item? is it 2? should we merge them?

Comment (krystofwoldrich) [reply src/js/client.ts:177]: It can happen, but if the server can handle it, why should we? 

Maybe to save a bit of space? Would that be a reason?

Comment (marandaneto) [reply src/js/client.ts:177]: It's cheaper to process on the server, has fewer data to transfer over the wire, and space in the disk, I recall folks from the ingest team also preferring a single client report per envelope whenever possible.

File: src/js/sdk.tsx (hunk 1 above)
Comment (marandaneto) [reply src/js/sdk.tsx:44]: So then this line does nothing?

File: src/js/client.ts (hunk 2 above)
Comment (krystofwoldrich) [reply src/js/client.ts:198]: Okay that's what I thought.

File: src/js/sdk.tsx (hunk 1 above)
Comment (krystofwoldrich) [reply src/js/sdk.tsx]: actually, we are not using the js default so it has to be there.
I got confused for a moment

File: src/js/client.ts [hunk 3]
```diff
@@ -116,8 +126,38 @@ export class ReactNativeClient extends BaseClient<ReactNativeClientOptions> {
...
+
+    let shouldClearOutcomesBuffer = true;
+    if (this._transport && this._dsn) {
+      this._transport.send(envelope)
+        .then(null, reason => {
+          logger.error('Error while sending event:', reason);
+          if (reason instanceof SentryError) { // SentryError is thrown by SyncPromise
+            shouldClearOutcomesBuffer = false;
+            // If this is called asynchronously we want the _outcomesBuffer to be cleared
+          }
```
Comment (marandaneto) [inline src/js/client.ts:147]: I'd have a logger call to the if and else branches.

**Review by marandaneto — APPROVED** [review]
Thanks @krystofwoldrich 
LGTM.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## Unreleased
+
+### Features
+
+- Add ClientReports ([#2496](https://github.com/getsentry/sentry-react-native/pull/2496))
```
Comment (marandaneto) [inline CHANGELOG.md:7]: I'd also add a note about the breaking change with self hosted versions, See the one on Android changelog file.

### Sentry Self-hosted Compatibility

- Starting with version `6.0.0` of the `sentry` package, [Sentry's self hosted version >= v21.9.0](https://github.com/getsentry/self-hosted/releases) is required or you have to manually disable sending client reports via the `sendClientReports` option. This only applies to self-hosted Sentry. If you are using [sentry.io](https://sentry.io), no action is needed.

File: src/js/sdk.tsx (hunk 1 above)
Comment (marandaneto) [inline src/js/sdk.tsx:44]: We have to sync this down to Android and iOS.

**Pull request merged**